core/flow/tc_warm_up.go

```diff
@@ -0,0 +1,104 @@
+package flow
+
+import (
+	"math"
+	"sync/atomic"
+
+	"github.com/alibaba/sentinel-golang/logging"
+
+	"github.com/alibaba/sentinel-golang/core/base"
+	"github.com/alibaba/sentinel-golang/core/config"
+	"github.com/alibaba/sentinel-golang/util"
+)
+
+type WarmUpTrafficShapingCalculator struct {
+	threshold         float64
+	warmUpPeriodInSec uint32
+	coldFactor        uint32
+	warningToken      uint64
+	maxToken          uint64
+	slope             float64
+	storedTokens      *uint64
+	lastFilledTime    *uint64
+}
+
+func NewWarmUpTrafficShapingCalculator(rule *FlowRule) *WarmUpTrafficShapingCalculator {
+	if rule.WarmUpColdFactor == 0 {
```

Comment:
The bad coldFactor value (<=0) should be filtered out when validating rules in flow rule manager (though checking is also needed here).